flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -0,0 +1,230 @@
+			// NOTE: This only do for the case where the operator is one-input operator. At present,
+			// any non-head operator on the operator chain is one-input operator.
+			synchronized (containingTask.getCheckpointLock()) {
+				endOperatorInput(1);
+			}
+		}
+
+		closeOperatorAndQuiesceTimeService(containingTask);
+
+		// transmit the close operation to the next operator on the operator chain in the topological order
+		if (transitive && next != null) {
+			next.close(containingTask, true, true);
+		}
+	}
+
+	private void closeOperatorAndQuiesceTimeService(StreamTask<?, ?> containingTask)
+		throws InterruptedException, ExecutionException {
+
+		final CompletableFuture<Void> closedFuture = new CompletableFuture<>();
+
+		// 1. executing the close operation must be deferred to the mailbox to ensure that mails already
+		//    in the mailbox are finished before closing the operator
+		// 2. to ensure that there is no longer output triggered by the timers before invoke the "endInput"
+		//    methods of downstream operators in the operator chain, we must quiesce the processing time
+		//    service to prevent the pending timers from firing, but allow the timers in running to finish
+		// 3. when the second step is finished, send a close mail to ensure that the timers yet in mailbox
+		//    are finished before exiting the following mailbox processing loop
+		// TODO: To ensure the strict semantics of "close", the operator should be allowed to decide how to
+		//  handle (cancel or trigger) the pending timers before being closed, and the second step should be
+		//  finished before the first step
+		deferCloseOperatorToMailbox(containingTask)
+			.thenRun(this::quiesceProcessingTimeService)
+			.thenRun(() -> sendClosedMail(closedFuture))
+			.exceptionally((throwable) -> forwardException(throwable, closedFuture));
+
+		// run the mailbox processing loop until the closing operation is finished
+		while (!closedFuture.isDone()) {
+			mailboxExecutor.yield();
+		}
+
+		// expose the exception thrown when closing
+		if (closedFuture.isCompletedExceptionally()) {
+			closedFuture.get();
+		}
+	}
+
+	private CompletableFuture<Void> deferCloseOperatorToMailbox(StreamTask<?, ?> containingTask) {
+		final CompletableFuture<Void> closeOperatorFuture = new CompletableFuture<>();
+
+		mailboxExecutor.execute(
+			() -> {
+				try {
+					closeOperator(containingTask);
+					closeOperatorFuture.complete(null);
+				} catch (Throwable t) {
+					closeOperatorFuture.completeExceptionally(t);
+				}
+			},
+			"StreamOperatorWrapper#closeOperator for " + wrapped
+		);
+		return closeOperatorFuture;
+	}
+
+	private void quiesceProcessingTimeService() throws CompletionException {
+		if (wrapped instanceof ProcessingTimeAware) {
+			ProcessingTimeService processingTimeService = ((ProcessingTimeAware) wrapped).getProcessingTimeService();
+			if (processingTimeService != null) {
+				checkState(processingTimeService instanceof ProcessingTimeServiceImpl);
```

Comment:
Operator is not required to have this specific `ProcessingTimeService` implementation, so we shouldn't throw an exception here.

Probably, the best way is to fix it is to pull `quiesce` method from `ProcessingTimeServiceImpl` to  `ProcessingTimeService` (it's not `@Public` and all current classes can implement it).